### DISCUSSION TITLE
How to create a trigger to update `auth.users.raw_user_meta_data`

### DISCUSSION THREAD

**Discussion by Falci** [supabase/supabase#20714] [Questions]
If I understood correctly, we can't make joins with `auth.users` ,

So I created a table in the public schema to mirror the `auth.user`.

```sql
CREATE TABLE IF NOT EXISTS "users" (
    id uuid PRIMARY KEY NOT NULL REFERENCES auth.users (id) ON DELETE CASCADE,
    name TEXT,
    avatar_url TEXT,
    created_at TIMESTAMP NOT NULL DEFAULT now(),
    updated_at TIMESTAMP
);
``` 

I'd like to use `auth.users` as "the source of truth", meaning: when the user updates its data, we change first `auth.users` (metadata) and that should be copied to `public.users`;

When a new user signs up, I need to check an external API to retrieve their name/avatar:

```sql
CREATE OR REPLACE FUNCTION public.handle_new_user()
RETURNS TRIGGER AS $$
DECLARE
  _user record;
BEGIN
  SELECT * FROM api.call_external_api(new.email) INTO _user; -- may return {"user_id":"123", "picture":"https://..."}

  IF _user.user_id IS NOT NULL THEN
    NEW.raw_user_meta_data = jsonb_build_object('name', _user.full_name, 'avatar_url', _user.picture)::jsonb;
    INSERT INTO public.users (id, name, avatar_url) VALUES (new.id, _user.full_name, _user.picture);
  ELSE
    NEW.raw_user_meta_data = jsonb_build_object('name', split_part(new.email, '@', 1))::jsonb;
    INSERT INTO public.users (id, name) VALUES (new.id, split_part(new.email, '@', 1));
  END IF;

  RETURN NEW;
END;
$$ LANGUAGE PLPGSQL
SECURITY DEFINER;

CREATE TRIGGER on_new_user AFTER INSERT ON auth.users FOR EACH ROW EXECUTE PROCEDURE public.handle_new_user ();
```

However, when a new user signs up, `raw_user_meta_data` is empty (`{}`).

What am I missing?

**Comment by Falci** [comment]
Ok, found the error:
We can't change properties of `NEW` when the trigger runs `AFTER INSERT/UPDATE`. It needs to be `BEFORE` otherwise won't make sense.

The solution was to split into 2 triggers: one before and one after (we can't create FK in the `before`).

However, there's an update on auth.users removing the metadata. I had to override it, which seems wrong. What should be the best way to keep it?

**Reply by GaryAustin1** [reply]
It definitely comes back and updates raw_app_meta_data after the insert.  I've observed that.
Based on my comment and link below I think it does the same with user metadata.

**Comment by GaryAustin1** [comment]
I THINK the issue is auth comes back and updates that column after the insert assuming it is not changing by the insert trigger.
https://github.com/supabase/gotrue/blob/0100a80aa2a0e22a4ce0f281079a900cd06c1df0/internal/models/user.go#L184
I know raw_app_meta_data is not updated until the update in the auth process.

Put an update trigger or modify your trigger to also take update.  Then use tg_op to decide which it is.
I would do a `raise log 'tg_op = %',tg_op;` and look in Postgres log to see what is happening.  You can also log metadata.
